git: use merged parent tree for git index

scott2000 opened the pull request:
I had some free time over the holidays, so I thought I'd take a shot at #3979. I think this matches what was agreed upon as the best "step 1" in the issue discussion.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

yuja commented:
Thanks for attacking this. I haven't reviewed the code, but we need to be aware that our tests don't cover most of Git index (or staging area) changes. For example, my last attempt to migrate `reset_head()` to `gix` had timestamp issue.
https://github.com/jj-vcs/jj/compare/main...yuja:jj:push-vpsmwmxpokyx

See https://github.com/jj-vcs/jj/issues/3786 for the timestamp issue.

We'll probably need to 1. add basic tests for Git index, 2. reimplement index updates by using gix, 3. switch to merged parent.

scott2000 commented:
> Thanks for attacking this. I haven't reviewed the code, but we need to be aware that our tests don't cover most of Git index (or staging area) changes. For example, my last attempt to migrate `reset_head()` to `gix` had timestamp issue. [main...yuja:jj:push-vpsmwmxpokyx](https://github.com/jj-vcs/jj/compare/main...yuja:jj:push-vpsmwmxpokyx)
>
> See #3786 for the timestamp issue.

Ahh I hadn't seen that. It seems more complicated than I had thought.

> We'll probably need to 1. add basic tests for Git index, 2. reimplement index updates by using gix, 3. switch to merged parent.

Makes sense to me.

scott2000 commented:
Ok I added tests for the current behavior with git2 and added a commit that migrates to gix without changing the behavior before the commit that changes it to use the merged parent tree.

This doesn't solve #3786, since we still use a "0:0" timestamp for newly added files, but I think it at least doesn't make it any worse than the old implementation (it still preserves timestamps for files which already had timestamps in the index).

I think it's a slight improvement actually, since previously we would lose all stat information when checking out a commit with conflicts (since the index would only contain the `.jjconflict` files), but now it will at least be preserved for files which aren't conflicted.

yuja commented on CHANGELOG.md:
In commit message:
> 1. If the working copy commit contains conflicted files which weren't
>   conflicted in the merged parent tree, then those files won't appear
>   as conflicts in the index (since the index is based on the parents).

In this case, I think user would have to do `git add` to commit the conflict by `git`. Is there any safety check in `git` that prevents accidental `git add`? (I'm just asking because I'm not familiar with the git behavior.)

scott2000 replied:
Yeah, that's correct; I think the safety check is only for `git commit`. Even though there's no risk of the user committing the conflict markers in that case without running `git add`, I still think we should add the dummy conflict. Mainly because if the user runs `git commit`, it will cause the unstaged files with conflict markers to be carried over to their new working copy, but the conflicted status won't carry over to the new working copy on the `jj` side, so they would end up with `@` having conflict markers without being conflicted.

yuja replied:
> Even though there's no risk of the user committing the conflict markers in that case without running `git add`, I still think we should add the dummy conflict. Mainly because if the user runs `git commit`, it will cause the unstaged files with conflict markers to be carried over to their new working copy, but the conflicted status won't carry over to the new working copy on the `jj` side,

True, but I don't think it's worth the complexity we have to scan both wc tree and parent tree. Maybe there would be other edge cases which can't be handled (something like user created new commit by `git commit`, then made conflicting state by using `git`, then ran `jj` which imports new wc-parent commit and conflicting wc)?

(BTW, if a committed tree had a file named `INDEX_DUMMY_CONFLICT_FILE`, maybe it should be indexed as a plain file.)

scott2000 replied:
Hmm yeah that's true. I'll change it.

yuja reviewed (commented):
Looks generally good, and the added tests are nice, thanks.

scott2000 commented on lib/src/git.rs:
```diff
@@ -1080,20 +1056,37 @@ pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(),
+        } else {
+            // We can't represent many-sided conflicts in the Git index, so just add the
+            // first side as staged even if there are conflicts.
+            push_index_entry(&path, entry.first(), Stage::Unconflicted);
```
Some IDEs (e.g. IntelliJ IDEA) use the index for their conflict resolution interfaces, and they ignore the materialized markers entirely. So if we put the first 3 as the conflict, then the user might accidentally resolve it using IntelliJ, which would lead to an incorrect resolution. So I think it's safer that we just don't add it as a conflict at all, since the user might think it's safe to use Git-based tools to resolve it if we do.

scott2000 commented on lib/src/git.rs:
```diff
@@ -1080,20 +1056,37 @@ pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(),
+    // same changes as `jj diff`. If the merged parent tree has 2-sided conflicts,
+    // then the Git index will also be conflicted.
+    for (path, entry) in wc_commit.parent_tree(mut_repo)?.entries() {
+        let entry = entry?.simplify();
```
> nit: it might be better to handle resolved case early. I'm not sure if it has measurable performance difference, though.

I'm not sure I understand what you mean by handling the resolved case early. Do you mean to remove the `.simplify()` somehow?

> Can you also check if the gix version doesn't regress the fast path [a075a5c](https://github.com/jj-vcs/jj/commit/a075a5c6cabd493759fc973d99d3fd88feb74735) ? (I assume it wouldn't, but just in case.)

Hmm it does look like it has a small regression when the tree didn't change (from ~580ms to ~615ms for `jj new` on an empty commit in nixpkgs on my machine). I did some profiling, and it looks like `TreeEntriesIterator::next` is 24% of the runtime of the command (75% of which is just reading the trees from the Git repo), `push_index_entry` is 2.6%, and writing the index to disk is 2.8%, leading to a total of 30% of the runtime of the command being spent in `reset_head`.

I'm not sure if it would be possible to reduce it by much, since 80% of the time in `reset_head` comes from iterating over the tree entries rather than anything to do with the index, and it seems like we'd have to iterate over the entries no matter what. Do you have any ideas? Or does a 6% regression seem acceptable for this case?

yuja replied:
> > nit: it might be better to handle resolved case early. I'm not sure if it has measurable performance difference, though.
> 
> I'm not sure I understand what you mean by handling the resolved case early. Do you mean to remove the `.simplify()` somehow?

Yes. I meant testing `.as_resolved()`/`.into_resolved()` first. It's probably better to handle resolved case differently from n-way conflicts (n > 3) anyway, for explicitness. Oh, btw, what would happen if n-way conflict were labeled as Ours (without Base and Theirs)? Does IntelliJ resolve it by taking the Ours?

> > Can you also check if the gix version doesn't regress the fast path [a075a5c](https://github.com/jj-vcs/jj/commit/a075a5c6cabd493759fc973d99d3fd88feb74735) ? (I assume it wouldn't, but just in case.)
> 
> Hmm it does look like it has a small regression when the tree didn't change (from ~580ms to ~615ms for `jj new` on an empty commit in nixpkgs on my machine).

Suppose nixpkgs has unusual structure (many directories compared to files), I think that's okay.

> For the non-fast-path case, it seems to have around the same performance as trunk for me.
> 
> I'm not sure if it would be possible to reduce it by much, since 80% of the time in `reset_head` comes from iterating over the tree entries rather than anything to do with the index, and it seems like we'd have to iterate over the entries no matter what. Do you have any ideas?

Maybe we could use gix API if the parent commit isn't a merge? I think gix one is faster compared to jj's native API (because of fewer abstraction layers.)

scott2000 replied:
I think I got it working with the `gix::Repository::index_from_tree` function when there's no conflict by copying the stat information afterwards. It seems to bring it down to about the same latency as the current trunk `jj` build for the fast path.

yuja commented on lib/tests/test_git.rs:
```diff
@@ -2184,21 +2196,184 @@ fn test_reset_head_with_index() {
+        .write_file(RepoPath::root(), &mut b"file".as_slice())
+        .block_on()
+        .unwrap();
+
```
nit: maybe we can use `testutils::write_*()` helpers

yuja commented on lib/src/git.rs:
```diff
@@ -987,73 +989,94 @@ fn update_git_head(
+    if mut_repo.git_head() != first_parent {
+        update_git_head(
+            &git_repo,
+            gix::refs::transaction::PreviousValue::Any,
```
nit: can you leave a TODO comment about #3754? Ideally we should ensure that the existing HEAD ref points to the known object id (but we have to somehow deal with non-detached state.)

fwiw, maybe we can use `PreviousValue::MustExist` for a little bit of safety.

yuja commented on lib/src/git.rs:
```diff
@@ -987,73 +989,94 @@ fn update_git_head(
+
+    let mut index = if &parent_tree_id == mut_repo.store().empty_tree_id() {
+        // If the tree is empty, gix fails to load the object, so we just use an empty
+        // index directly.
```
nit: "gix fails" -> "gix can fail"?

It might be our fault that we don't write an empty tree object to the store even if we create a real Git commit with an empty tree.

yuja commented on lib/src/git.rs:
```diff
@@ -1022,38 +1026,28 @@ pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(),
+    // Use the merged parent tree as the Git index, allowing `git diff` to show the
+    // same changes as `jj diff`. If the merged parent tree has 2-sided conflicts,
+    // then the Git index will also be conflicted.
+    let mut index = if let Some(tree) = parent_tree.as_merge().as_resolved() {
```
nit: could you add lib test that covers both resolved and merged cases?

yuja commented on lib/src/git.rs:
```diff
@@ -1080,6 +1074,106 @@ pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(),
+                }
+            }
+            TreeValue::Symlink(id) => (id.as_bytes(), gix::index::entry::Mode::SYMLINK),
+            TreeValue::Tree(tree_id) => {
+                for (path, entry) in store.get_tree(path.to_owned(), tree_id)?.entries() {
+                    push_index_entry(index, store, &path, &Some(entry), stage)?;
+                }
+                return Ok(());
+            }
```
This seems odd because `.entries()` is a recursive iterator.

I think it handles the case where tree conflicts with non-tree object. In that case, we would probably treat the path as non-tree. I don't remember what would be materialized in the working copy, but I don't think it's correct to recursively insert Git index entries.

scott2000 replied:
Yeah, I added this to handle the case of conflicts between tree and non-tree. When we materialize it, it looks like this (for conflict between "test" and "test/file"):

```
Conflict:
  Adding file with id f73f3093ff865c514c6c51f867e35f693487d0d3
  Adding tree with id c3e9a16324d8ebd732b8a6aa79fa6881238c4ea6
```

Currently, I add all of the files in the tree to the index (which is why it needs to be recursive), and it produces this `git status` output:

```
Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
        added by us:     test
        added by them:   test/file
```

But it looks like Git does something different in this case; `git merge` keeps the tree and renames the file to something else so that it doesn't conflict with the name of the tree, so in Git the merge would produce something like:

```
You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:
        deleted:    test
        new file:   test/file

Unmerged paths:
  (use "git add <file>..." to mark resolution)
        added by us:     test~HEAD
```

Since this is an uncommon case, maybe we could just skip adding the tree to the index if this case happens? Then it would look like this:

```
Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
        added by us:     test
```

What do you think?

martinvonz replied:
> When we materialize it, it looks like this (for conflict between "test" and "test/file"):

That gets written as a file called `test`, right? If we write a file like that to the working copy, then I think we should also just write a single matching entry to the git index. Does that make sense?

scott2000 replied:
> That gets written as a file called `test`, right?

Yes, that's correct.

> If we write a file like that to the working copy, then I think we should also just write a single matching entry to the git index. Does that make sense?

That makes sense to me. I think we can just skip adding the tree then, so only `test` would be added to the index.

yuja reviewed (approved):
LG, thanks!
